Timeouts & TTLs

mattheath opened the pull request:
Very much a work in progress, but attempts to put a deadline on all requests to Zookeeper. This will hopefully fix a number of suspected problems where calls to the library might never return.
- [x] Switch logging to `seelog`, rather than `log` so we have more control over log levels
- [x] Allow a custom request timeout to be set on a connection, defaults to `3s` which is probably far too high.
- [x] Abandon requests which exceed the request timeout - applies to all requests
- [x] Split apart timeouts and TTLs on Locks, allowing these to be set independently
- [x] Return `ErrRequestTimeout` on request timeout

Unrelated, but used by the Lock implementation
- [x] Add `Reconnect()` method which kills the ZK connection forcing a reconnect _...something something danger zone_

For the moment I'm not going to implement per request timeouts, as it's going to take me too long. One way would be a background worker running off a queue ordered by TTL (eg. [a priority queue](http://en.wikipedia.org/wiki/Priority_queue), [example](http://golang.org/pkg/container/heap/#example)) which on TTL expiry of a request fired a timeout error down the response channel.

mattheath committed:
Split out timeout and lock TTL into separate things

mattheath committed:
Switch logging to seelog, so we can use log levels

mattheath committed:
Allow a request timeout to be set on the connection

mattheath committed:
Timeout requests that take too long

mattheath committed:
Return correct zxid on timeout error case

mattheath committed:
Update queueRequest to return the request xid, so we can clean up

mattheath committed:
Remove timed out requests from the request map

domwong commented on zk/conn.go:
```diff
@@ -322,7 +336,8 @@ func (c *Conn) sendSetWatches() {
 		if err != nil {
-			log.Fatal(err)
+			log.Criticalf("[Zookeeper] %s", err.Error())
+			panic(err)
```
Panic on error? Or do we want to try and recover?

mattheath replied:
This is terrible, and we should definitely recover. Its matching the previous behaviour in the library that called `log.Fatal` which logs and exits with `os.Exit(1)` just more explicit as seelog doesn't provide this (for obvious reasons!)

mattheath replied:
I've added a `Reconnect()` method which we can use to kill the underlying `net.Conn` forcing a reconnect (we already do this in `Lock` though didn't previously expose it). So I think we have two choices:
- Panic. _I'm not keen on this..._ I don't think the library should ever actually panic, though without understanding more about ZK I'm not sure how _critical_ this error is?
- `Reconnect()` which will kill the network connection to Zookeeper, causing the send and recv channels to exit, and all existing requests to be flushed. We would then reconnect, call `authenticate()` and get back to here, so we could recover, or we could end up in a pretty tight connection loop. Could add a delay before doing this I guess?

domwong commented on zk/conn.go:
```diff
@@ -591,12 +610,30 @@ func (c *Conn) queueRequest(opcode int32, req interface{}, res interface{}, recv
+		// Request timed out, clean up
+		log.Warnf("[Zookeeper] Request timed out after %v", c.requestTimeout)
+
+		// Delete the outstanding request from the map, we will then ignore any response
```
What happens if we have a request that just takes a long time and returns having correctly created the node (or whatever). Will applications have to handle this scenario any differently? (just thinking out loud)

mattheath replied:
Yeah this is definitely a problem, though arguably it could already happen. Not sure on the best way to solve this, but I would assume the application would have to handle this in an error scenario, by reattempting and checking for existence of nodes.

The Lock mechanism already does something similar to this - it uses a guid when creating an ephemeral sequential node that is used across requests, so it can identify if it succeeded but got an error response.

mattheath replied:
Added `ErrRequestTimeout` as an error type which can be caught in this case. The client should then handle this, by either retrying, or checking if the previous operation was successful.

mattheath committed:
Return timeout error when timing out

mattheath committed:
Expose CloseConnection which forces a reconnect

This is done by killing the underlying net connection, which causes both send
and recv loops to exit, flush requests and reconnect

mattheath committed:
Make Reconnect method name more obvious
